determined-ai/determined
# feat: Enable viewing epoch based metric data in UI [WEB-859]

## Description

Adds the ability to view `epoch` based training data in the WebUI. 

We want to allow users to view metric data based on epochs via metrics reported to the `CoreAPI` in the WebUi. For now, we will expect users to define an `epoch` metric to be reported to `CoreAPI` and this metric will be used as an `x-axis` label in the WebUI. Once the updated chart features are ready to ship. We will add documentation to explain to users how they can record and view `epoch` based training data. 

Part of this PR includes backend updates regarding how metric data is returned in the API layer. The updates were requested in an effort to better support arbitrary x-axis in the future. For now this PR introduces a hardcoded `epoch` metric label throughout the front and backend. However, the updates have been made with the idea that in the future we will support dynamic x-axis values and more than a hardcoded `epoch` value.  For example an `xAxisMetricLabels` array has been introduced to simulate handling a dynamic number of potential x axis arguments. 




## Test Plan

1. Ensure the charts in the `overview` for both single and multi trials experiments look the same as it does today. There should be no differences in functionality for existing chart behavior. 
2. Report training data that includes `epoch` as a metric to the CoreAPI. Docs for reference: https://docs.determined.ai/latest/reference/reference-training/training/api-core-reference.html

to accomplish this you can perform the following steps 

- Update `examples/tutorials/core_api/1_metrics.py` to report `epoch`, `training_accuracy`, and `validation_accuracy`
  you can use the following code
  
```
import logging
import sys
import time
import determined as det

def main(core_context):
    for batch in range(0,100):
        training_accuracy =  1 - (10/(batch+11))
        validation_accuracy = training_accuracy - 0.05
        steps_completed = batch + 1
        time.sleep(0.1)
        # NEW: report training metrics.
        core_context.train.report_training_metrics(
            steps_completed=steps_completed, metrics={"epoch": batch, "training_accuracy":training_accuracy}
        )
        core_context.train.report_validation_metrics(steps_completed=steps_completed, metrics={"epoch": batch, "validation_accuracy":validation_accuracy})

if __name__ == "__main__":
    # NEW: enable logging, using the det.LOG_FORMAT.  Enabling
    # logging enables useful log messages from the determined library,
    # and det.LOG_FORMAT enables filter-by-level in the WebUI.
    logging.basicConfig(level=logging.DEBUG, format=det.LOG_FORMAT)
    # Log at different levels to demonstrate filter-by-level in the WebUI.
    logging.debug("debug-level message")
    logging.info("info-level message")
    logging.warning("warning-level message")
    logging.error("error-level message")

    with det.core.init() as core_context:
        main(core_context=core_context)
  ```
  
  - in the terminal navigate to `examples/tutorials/core_api`
  - run the command `det e create -f 1_metrics.yaml .` to start an experiment 
  
  3.  Navigate to the experiment created in (2) and ensure that only the defined metrics are shown. For example only `training_accuracy` and `validation_accuracy`.
  
  
![Screen Shot 2023-02-07 at 6 22 26 PM](https://user-images.githubusercontent.com/103522725/217397080-0ffe05a7-f5fd-44f8-b722-b4f0afefcaae.png)

In the `Metrics` dropdown ensure that the only options are `All`, `validation_accuracy`, and `training_accuracy` or whichever metrics were defined. 

4. Turn on the chart feature (use `&f_chart=on`) and view the same experiment.

5. Ensure that `Epoch` is in the `X-Axis` options. 

![Screen Shot 2023-02-07 at 6 26 05 PM](https://user-images.githubusercontent.com/103522725/217397499-03fcd34e-9c25-4f9f-ad30-9c62e80921b9.png)

6. Ensure that the graphs for all of the `X-Axis` options `Batches`, `Epoch`, and `Time` vs `Accuracy` all contain valid data.  

![Screen Shot 2023-02-07 at 6 27 41 PM](https://user-images.githubusercontent.com/103522725/217397676-75b27304-2269-4703-ae23-0829f3f9bf95.png)

![Screen Shot 2023-02-07 at 6 28 01 PM](https://user-images.githubusercontent.com/103522725/217397716-1598a000-24b5-453f-964a-cb63037cc413.png)

![Screen Shot 2023-02-07 at 6 35 53 PM](https://user-images.githubusercontent.com/103522725/217398723-b4d25983-b354-4ce3-a9c4-9371d7cc1396.png)

![Screen Shot 2023-02-07 at 6 28 14 PM](https://user-images.githubusercontent.com/103522725/217397741-83872b39-251c-4279-b35d-8e0d56791fa0.png)

![Screen Shot 2023-02-07 at 6 27 30 PM](https://user-images.githubusercontent.com/103522725/217397657-287b9ad8-9081-481c-b3ea-6e234c4905ef.png)

7. Ensure that epoch is not shown as a `y-axis` metric. There should not be any `epoch` vs `time`,`batches', or `epoch` graph.
8. View any experiment that does not contain `epoch` data. And Ensure the `chart` feature is on. 
9. Ensure that `Epoch` is not an `X-Axis` option. 

![Screen Shot 2023-02-07 at 6 30 02 PM](https://user-images.githubusercontent.com/103522725/217397969-ef0eb2e3-8cf0-43ac-a46e-d0c7344a0708.png)





## Commentary (optional)





## Checklist

~- [ ] User-facing API changes need the "User-facing API Change" label.~
~- [ ] Release notes should be added as a separate file under `docs/release-notes/`.~
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
~- [ ] Licenses should be included for new code which was copied and/or modified from any external code.~

## Ticket

WEB-859
